src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs

```diff
@@ -21,15 +21,16 @@ partial struct HandleCallAction
 		static ValueSetLattice<SingleValue> MultiValueLattice => default;
 
 		readonly DiagnosticContext _diagnosticContext;
+		readonly AnnotationContext _annotationContext;
 		readonly RequireDynamicallyAccessedMembersAction _requireDynamicallyAccessedMembersAction;
 
 		public bool Invoke (MethodProxy calledMethod, MultiValue instanceValue, IReadOnlyList<MultiValue> argumentValues, out MultiValue methodReturnValue, out IntrinsicId intrinsicId)
 		{
 			MultiValue? returnValue = null;
 
 			bool requiresDataFlowAnalysis = MethodRequiresDataFlowAnalysis (calledMethod);
-			DynamicallyAccessedMemberTypes returnValueDynamicallyAccessedMemberTypes = requiresDataFlowAnalysis ?
-				GetReturnValueAnnotation (calledMethod) : 0;
+			var annotatedMethodReturnValue = _annotationContext.GetMethodReturnValue (calledMethod);
+			Debug.Assert (requiresDataFlowAnalysis || annotatedMethodReturnValue.DynamicallyAccessedMemberTypes == DynamicallyAccessedMemberTypes.None);
```

Comment:
Nit: we could inline `MethodRequiresDataFlowAnalysis(calledMethod)` and remove `requiresDataFlowAnalysis` since it's only used in one place outside of the assert.

Reply:
I can't inline it exactly because it's called from the shared code - but I moved it to `AnnotationContext` (now `FlowAnnotations` in response to your other suggestion.)